core/src/main/java/dk/alexandra/fresco/framework/util/DeterministicSecureRandom.java

```diff
@@ -0,0 +1,37 @@
+package dk.alexandra.fresco.framework.util;
+
+/**
+ * <p>
+ * Java's SecureRandom is not 'deterministic' in the sense that calls to setSeed (at least using
+ * some crypto providers) only adds to the initial seed, that is taken from system state when
+ * SecureRandom was created.
+ * </p>
+ * <p>
+ * Often, in our protocols, we need a SecureRandom that can be seeded 'determinitically', e.g. in
+ * the sense that two instances created with the same seed yields the same sequence of random bytes,
+ * but can still not be easily guessed by an adversary.
+ * </p>
+ * <p>
+ * Implementations of this class will have the property that data generated is pseudorandom and if
+ * parties calls {@link #setSeed(byte[])} with the same argument, subsequent calls to
+ * {@link #nextBytes(byte[])} will be deterministic and secure. For actual security guarantees, we
+ * refer to the individual implementations.
+ * </p>
+ */
+public interface DeterministicSecureRandom {
+
+  /**
+   * Has the same functionality as {@link java.security.SecureRandom#nextBytes(byte[])}
+   * 
+   * @param bytes The byte array which will be overwritten with random data.
+   */
+  public void nextBytes(byte[] bytes);
+
+  /**
+   * Sets the seed used for generating randomness. If parties need to agree on the randomness
+   * output, then the seed must be identical for all parties.
+   * 
+   * @param seed The initial randomness.
+   */
+  public void setSeed(byte[] seed);
```

Comment:
The random generator should not expose seed in its interface.

Reply:
Yes it should. You have to be able to reset the generator with a new seed. This should be done approx every 2^42 times you call getBytes. 
You could alternatively create a new instance, but the NIST standard requires this to be available.

Reply:
It should be reset'able, but not by the persons getting the random bits. It can even be handled internally in the class producing the psuedorandomness.

You are making the interface defining two sets of responsibilities. The reset is internal to the users of the randomness.

Reply:
No it cannot be handled internally within the class. It needs external randomness to be secure. 
I still believe the current approach is the correct one. Who would reset it if not the party using the generator? The reset has to happen at the exact same moment in code (i.e. no race conditions). 
The perfect way is maybe to create listeners which are triggered when limit is reached, and a Semaphore or similar lock preventing users from obtaining more before the reset is done. Who would listen? Well, the parties using the generator (i.e. the protocol suite) would. 
It seems too much of a hassle without actually helping.

Reply:
See other comment